Workflows/Environment Synchronization.md

```diff
@@ -0,0 +1,126 @@
+    remote_db_dump_file_path: './tmp/downloaded_db_dump.sql'
+  run_migrations_on_local_db:
+  delete_remote_db_dump:
+    remote_db_dump_file_path: './tmp/downloaded_db_dump.sql'
+  execute_custom_command:
+    command: "RAILS_ENV=staging bundle exec rake \"remont[db/s3_sync.rb]\""
+```
+
+Not all steps are mandatory. For example, you may have obtained the database dump in some other way,
+i.e. not by downloading it from the S3 bucket, or you may not want to delete the downloaded DB dump file.
+The option of adding a final custom command can be used to run a rake task for 
+file synchronization or any other shell command.
+
+This gem has also been added to the JSONAPI Example App.
+
+### File synchronization
+File synchronization is an optional step that not everyone may need when syncing environments,
+but it can also be a standalone task that is run whenever necessary.
+
+In addition to that, there could be very different requirements as to what files should be synchronized and how.
+For example, you may want to sync entire AWS S3 buckets, or you may want to just copy a couple of files from one
+bucket to another, or a file may have several different versions and you may want to handle that in a particular way.
+You may want to use the AWS CLI tool or the [aws-sdk-ruby gem](https://github.com/aws/aws-sdk-ruby)
+with options specific to your use case.
+
+Before you decide on the implementation, take some time to consider if syncing files is even a good idea.
+For instance, if you want to sync entire buckets, keep in mind that the size of the files you are copying to
+the destination bucket will affect your AWS costs. Also, if you are copying files from the production environment,
+they may contain sensitive data, such as proprietary and/or personally identifiable information that you cannot easily
+anonymize.
+
+If you decide to proceed with file synchronization, you can choose to implement the solution we have included in
+the JSONAPI Example App. It contains a template (the `s3_sync.rb` script with accompanying classes)
+for copying files from one bucket to another for a specified set of records in your database.
+The remont gem is a flexible tool that has been reused here and the corresponding rake task can be added as
+the final step to the data sync step definitions required by the copy_bot gem or it can be run separately.
+
+The proposed solution uses the remont gem DSL to iterate over records for which you want to copy files from one bucket
+to another. For each record the `EnvSync::FileSync::Processor` is called, and this processor calls
+the `EnvSync::FileSync::Synchronizer`. You will notice that in the `EnvSync::FileSync::Synchronizer`
+we are just copying one file from one bucket to another. However, you can use this example and adjust it to you needs
+by simply changing the code in the `copy_file`method or by adding a completely different method, just be sure
+to update the `EnvSync::FileSync::Processor` if necessary.
```

Comment:
Where are the production (or any other source env) s3 secrets saved for this synchronization?

https://github.com/infinum/rails-jsonapi-example-app/blob/9f5505211435cf4a693aeda780d806d8219b8fa9/app/domains/env_sync/file_sync/synchronizer.rb#L5

I can see here that you have "source" secrets listed, but there's not way to distinguish them between local secrets (e.g. with a prefix of some kind 🤷‍♀️ )